synapse/app/generic_worker.py

```diff
@@ -206,10 +206,15 @@ async def on_POST(self, request, device_id):
 
         if body:
             # They're actually trying to upload something, proxy to main synapse.
-            # Pass through the auth headers, if any, in case the access token
-            # is there.
-            auth_headers = request.requestHeaders.getRawHeaders(b"Authorization", [])
-            headers = {"Authorization": auth_headers}
+
+            # Proxy headers from the original request, such as the auth headers
+            # (in case the access token is there) and the original IP /
+            # User-Agent of the request.
+            headers = {
+                header: request.requestHeaders.getRawHeaders(header, [])
+                for header in (b"Authorization", b"X-Forwarded-For", b"User-Agent")
```

Comment:
It's worth noting that X-Forwarded-For is a *list* of servers that the request has passed through. Technically speaking, we ought to add the IP address that we received the request from (ie, `request.client.host`) to the end of the list.

On the other hand, given that Synapse ignores all but the first element of the list, and I think it's impossible to get here without having passed through at least one proxy which will add an `X-Forwarded-For` header, it's somewhat moot currently.

Reply:
> Technically speaking, we ought to add the IP address that we received the request from (ie, `request.client.host`) to the end of the list.

I'm not 100% sure I understand this suggestion. Shouldn't the client's IP already be in the `X-Forwarded-For` if we went through a proxy? Or are you suggesting that if we don't find an `X-Forwarded-For` we should add one that is the client IP? It seems like we should be adding the worker's IP as the last IP in the list, not the client IP?

Reply:
Basically, each time the request passes through a proxy:
 * if it's the first proxy in the chain, the proxy should add an X-Forwarded-For header with the IP address of the client
 * otherwise, the proxy should add the IP of the *previous* proxy to the list.

So the server at the far end can see the whole chain.

> Shouldn't the client's IP already be in the X-Forwarded-For if we went through a proxy? 

the *client's* IP should be, but not that of the proxy just before synapse (haproxy, in our case). I'm suggesting we add haproxy's IP address to the list: ie, `request.client.host`.

> Or are you suggesting that if we don't find an `X-Forwarded-For` we should add one that is the client IP?

If there is *no* `X-Forwarded-For`, that would mean we are the first server in the chain, so yes we would add one that is the client IP:  ie, `request.client.host`.

> It seems like we should be adding the worker's IP as the last IP in the list, not the client IP?

we don't need to add the worker's IP at all.

Reply:
something like this, maybe:

```python
xff = request.requestHeaders.getRawHeaders(b"X-Forwarded-For", "")
if isinstance(request.client, (address.IPv4Address, address.IPv6Address)):
    if xff:
        xff += ", "
    xff += request.client.host
headers["X-Forwarded-For"] = xff
```

Reply:
Yeah, I see. I was missing that you add the previous server in the chain, which is `request.client.host`. 👍 (For some reason I was thinking that was the true client IP...)

Reply:
I think I've added this using similar code to the above.